### PR TITLE
SQL Task Dmitry Gaiduk

### PR DESCRIPTION
**Pull request by IdokkeI**

### CODE REVIEW

**Comment by IdokkeI** [comment]
https://travis-ci.com/github/IdokkeI/db-assignments/builds/181915883

File: task/sql-tasks.js [hunk 1]
```diff
@@ -105,7 +154,17 @@ async function task_1_6(db) {
...
 async function task_1_7(db) {
-    throw new Error("Not implemented");
+    let result = await db.query(`
+    select 
+        e2.EmployeeId as "EmployeeId",
+        concat( e2.FirstName, " ", e2.LastName) as "FullName",
+        coalesce(concat( e1.FirstName, " ", e1.LastName), '-') as "ReportsTo"
+    from 
+       Employees e1
+           right join Employees e2 on e1.EmployeeID = e2.ReportsTo
```
Comment (AlexMokin) [inline task/sql-tasks.js]: Don't use right join if not necessary

File: task/sql-tasks.js [hunk 2]
```diff
@@ -177,7 +274,13 @@ async function task_1_12(db) {
  *
  */
 async function task_1_13(db) {
-    throw new Error("Not implemented");
+    let result = await db.query(`
+    SELECT
```
Comment (AlexMokin) [inline task/sql-tasks.js:277]: 1) use count when you need to count recrods to show intent explicitly

2) it's better to write this query like this

```
select
(select count()...) as 'T'
(select count()...) as 'TT'
```

notice from-clause is optional

File: task/sql-tasks.js [hunk 3]
```diff
@@ -129,7 +197,14 @@ async function task_1_8(db) {
  *
  */
 async function task_1_9(db) {
-    throw new Error("Not implemented");
+    let result = await db.query(`
+    SELECT 
+	CustomerID,
+    ContactName
+ FROM Customers
+ where ContactName like 'F%___n%'
```
Comment (AlexMokin) [inline task/sql-tasks.js]: n must be exatly 4rth character

File: task/sql-tasks.js [hunk 4]
```diff
@@ -199,7 +310,23 @@ async function task_1_14(db) {
...
+(select count(*) from Orders where year(OrderDate) = 1997 and month(OrderDate) = 4) as "April",
+(select count(*) from Orders where year(OrderDate) = 1997 and month(OrderDate) = 5) as "May",
+(select count(*) from Orders where year(OrderDate) = 1997 and month(OrderDate) = 6) as "June",
+(select count(*) from Orders where year(OrderDate) = 1997 and month(OrderDate) = 7) as "July",
+(select count(*) from Orders where year(OrderDate) = 1997 and month(OrderDate) = 8) as "August",
+(select count(*) from Orders where year(OrderDate) = 1997 and month(OrderDate) = 9) as "September",
+(select count(*) from Orders where year(OrderDate) = 1997 and month(OrderDate) = 10) as "October",
+(select count(*) from Orders where year(OrderDate) = 1997 and month(OrderDate) = 11) as "November",
+(select count(*) from Orders where year(OrderDate) = 1997 and month(OrderDate) = 12) as "December"
+from Orders group by 1,2,3,4,5,6,7,8,9,10,11,12
```
Comment (AlexMokin) [inline task/sql-tasks.js]: using group by doesn't make sense, if you do it like this you coud've just skip from-clause altogether. from clause is optional

Comment (AlexMokin) [reply task/sql-tasks.js]: try to change this query to look like this 

select 
   count(<some expression) as "JAN",
   ......
FROM Orders
where year = 1997

File: task/sql-tasks.js [hunk 5]
```diff
@@ -223,7 +358,15 @@ async function task_1_16(db) {
  *
  */
 async function task_1_17(db) {
-    throw new Error("Not implemented");
+    let result = await db.query(`
+    SELECT c.CategoryName,
+		avg(p.UnitPrice) as "AvgPrice"
+FROM Products p
+	join Categories c on c.CategoryID = p.CategoryID
```
Comment (AlexMokin) [inline task/sql-tasks.js]: use inner join explicitly. fix this in the whole file not just in this task

File: task/sql-tasks.js [hunk 6]
```diff
@@ -280,7 +462,27 @@ async function task_1_21(db) {
...
+        FROM Customers 
+        JOIN Orders ON Customers.CustomerID=Orders.CustomerID
+        JOIN OrderDetails ON Orders.OrderID=OrderDetails.OrderID
+        JOIN Products ON OrderDetails.ProductID=Products.ProductID
+        WHERE OrderDetails.UnitPrice = (SELECT
+                    MAX(od.UnitPrice)
+                FROM Customers as c
+                JOIN Orders as o ON o.CustomerID=c.CustomerID
+                JOIN OrderDetails as od ON od.OrderID=o.OrderID
+                WHERE Customers.CompanyName=c.CompanyName
```
Comment (AlexMokin) [inline task/sql-tasks.js]: company name doesn't uniquely identify customer. also this column doesn't have an index so it's also slow.

Comment (AlexMokin) [reply task/sql-tasks.js]: also try to use group by customer id and join subquery on unit price and customer id to speed up query

**Review by AlexMokin — COMMENTED** [review]
Looks good. See comments. Also format code according to this code style https://gist.github.com/fredbenenson/7bb92718e19138c20591

I'll mark the PR as "Draft", please click "ready for review" when it will be finished. Thank you!

File: task/sql-tasks.js (hunk 3 above)
Comment (IdokkeI) [reply task/sql-tasks.js:205]: Yes i know but if i didn't use % after F i didn't get result which i need.

**Comment by IdokkeI** [comment]
https://travis-ci.com/github/IdokkeI/db-assignments

**Comment by IdokkeI** [comment]
can you check my ui, please https://idokkei.github.io/skype/

**Comment by IdokkeI** [comment]
https://travis-ci.com/github/IdokkeI/db-assignments/builds/183870989

File: task/sql-tasks.js [hunk 7]
```diff
@@ -105,7 +154,17 @@ async function task_1_6(db) {
...
 async function task_1_7(db) {
-    throw new Error("Not implemented");
+    let result = await db.query(`
+    select 
+        e2.EmployeeId as "EmployeeId",
+        concat( e2.FirstName, " ", e2.LastName) as "FullName",
+        ifnull(concat( e1.FirstName, " ", e1.LastName), '-')  as "ReportsTo"
+    from 
+       Employees e1
+           right join Employees e2 on e1.EmployeeID = e2.ReportsTo
```
Comment (AlexMokin) [inline task/sql-tasks.js]: Don't use right join if not necessary

File: task/sql-tasks.js (hunk 2 above)
Comment (AlexMokin) [inline task/sql-tasks.js:277]: 1) use count when you need to count recrods to show intent explicitly

2) it's better to write this query like this

```
select
(select count()...) as 'T'
(select count()...) as 'TT'
```

notice from-clause is optional

File: task/sql-tasks.js [hunk 8]
```diff
@@ -199,7 +310,24 @@ async function task_1_14(db) {
  *
  */
 async function task_1_15(db) {
-    throw new Error("Not implemented");
+    let result = await db.query(`
+    select 
+        sum(month(OrderDate) = 1)  as "January",
```
Comment (AlexMokin) [inline task/sql-tasks.js]: use count

File: task/sql-tasks.js [hunk 9]
```diff
@@ -280,7 +463,27 @@ async function task_1_21(db) {
...
+    select
+	distinct
+            Customers.CompanyName,
+            Products.ProductName,
+            OrderDetails.UnitPrice as "PricePerItem"
+        from Customers 
+        inner join Orders ON Customers.CustomerID=Orders.CustomerID
+        inner join OrderDetails ON Orders.OrderID=OrderDetails.OrderID
+        inner join Products ON OrderDetails.ProductID=Products.ProductID
+        WHERE OrderDetails.UnitPrice = (select
```
Comment (AlexMokin) [inline task/sql-tasks.js]: company name doesn't uniquely identify customer. also this column doesn't have an index so it's also slow.
also try to use group by customer id and join subquery on unit price and customer id to speed up query

**Review by AlexMokin — COMMENTED** [review]
See comments. Also please format code according to this code style https://gist.github.com/fredbenenson/7bb92718e19138c20591

I'll mark the PR as "Draft", please click "ready for review" when it will be finished. Thank you!

File: task/sql-tasks.js (hunk 8 above)
Comment (IdokkeI) [reply task/sql-tasks.js:315]: but if i'll use count i'll get the quantity all records

**Comment by IdokkeI** [comment]
https://travis-ci.org/github/IdokkeI/db-assignments/builds/726544584

**Comment by IlyaMokin** [comment]
Closing the PR as accepted, but please be more careful in the future in terms of redundant changes. I see a lot of edits in comments which maybe can be useful as separate PR to the project but redundant for the task.

**Pull request closed without merging**